source/adapters/level_zero/kernel.cpp

```diff
@@ -191,34 +191,62 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueKernelLaunch(
   ZE2UR_CALL(zeKernelSetGroupSize, (ZeKernel, WG[0], WG[1], WG[2]));
 
   bool UseCopyEngine = false;
+  bool LastCommandEventIncluded = false;
   _ur_ze_event_list_t TmpWaitList;
   UR_CALL(TmpWaitList.createAndRetainUrZeEventList(
-      NumEventsInWaitList, EventWaitList, Queue, UseCopyEngine));
+      NumEventsInWaitList, EventWaitList, Queue, UseCopyEngine,
+      &LastCommandEventIncluded));
 
   // Get a new command list to be used on this call
   ur_command_list_ptr_t CommandList{};
   UR_CALL(Queue->Context->getAvailableCommandList(
       Queue, CommandList, UseCopyEngine, true /* AllowBatching */));
 
+  uint32_t WaitListLength = TmpWaitList.Length;
+  uint32_t WaitEventOffset = 0;
+
+  if (Queue->Device->useDriverInOrderLists() && Queue->isInOrderQueue()) {
+    LastCommandEventIncluded &=
+        Queue->LastUsedCommandList != Queue->CommandListMap.end() &&
+        CommandList->first != Queue->LastUsedCommandList->first;
+
+    if (!LastCommandEventIncluded && WaitListLength > 0) {
+      WaitListLength--;
+      WaitEventOffset = 1;
+      if (Queue->LastCommandEvent && Queue->LastCommandEvent->ZeEvent) {
+        ZE2UR_CALL(zeEventDestroy, (Queue->LastCommandEvent->ZeEvent));
```

Comment:
why event is getting destroyed here?
what about ref counts?
it may be in use by some other calls